### PR TITLE
Discord Scheduler: allow role mentions

### PR DESCRIPTION
**Pull request by rshipp**
### Description

Allow mentioning roles by name, instead of only `@everyone`.

### Motivation and Context

Task: https://app.asana.com/0/1201131148207877/1201314155101891/f

### Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

### Checklist:

- [ ] My code follows the code style of this project.
- [ ] I have added neccessary labels
- [ ] My PR title follows the convention in the guide
- [ ] I have assigned myself to the PR
- [ ] I have requested reviews from team members
- [ ] I have created a task for reviewers on PM tool
- [ ] All new and existing tests passed.

### CODE REVIEW

File: spec/services/schedule_discord_message_spec.rb [hunk 1]
```diff
@@ -5,7 +5,7 @@
 require 'discord_bot'
 
 RSpec.describe ScheduleDiscordMessage do
-  describe '#run' do
+  describe '.run' do
```
Comment (elenabaurkot) [inline spec/services/schedule_discord_message_spec.rb:8]: What is the difference between using describe '#run' vs '.run'?

**Review by elenabaurkot — COMMENTED** [review]
Lgtm!

File: spec/services/schedule_discord_message_spec.rb (hunk 1 above)
Comment (rshipp) [reply spec/services/schedule_discord_message_spec.rb:8]: `#blah` is the syntax used for controller actions, like `#new`. i think we just have it copied in other places in our specs by accident. these names are just by convention though, so it doesn't break anything either way.

File: app/helpers/discord_schedule_helper.rb [hunk 2]
```diff
@@ -0,0 +1,18 @@
...
+  ROLE_MENTION_REGEX = /<@&(\d+)>/
+
+  def render_discord_message(message)
+    # For now, this is just the inverse of
+    # ScheduleDiscordMessage.convert_role_mentions.
+    message.scan(ROLE_MENTION_REGEX).flatten.each do |discord_role_id|
+      # This is a role mention.
+      role = DiscordServerRole.find_by(discord_role_id: discord_role_id)
+      # Skip invalid role IDs, instead of crashing.
+      next unless role
```
Comment (sadleb) [inline app/helpers/discord_schedule_helper.rb:11]: Worth adding a Honeycomb field?

File: app/services/schedule_discord_message.rb [hunk 3]
```diff
@@ -35,10 +36,28 @@ def run
...
 
       # Schedule the job.
       SendDiscordMessageJob.set(wait_until: run_at).perform_later(server_id, channel_name, message_content)
     end
   end
+
+  def self.convert_role_mentions(server, message)
+    if message.include? '@'
+      # Possible role mention(s).
+      server.roles.sort_by { |r| -r.name.length }.each do |role|
```
Comment (sadleb) [inline app/services/schedule_discord_message.rb:49]: This seems fine, but I had to look up `sort_by` usage to understand the `-r` syntax. Another option is `server.roles.sort_by { |r| r.name.length }.reverse` which may be more readable?

File: app/services/schedule_discord_message.rb [hunk 4]
```diff
@@ -35,10 +36,28 @@ def run
...
+        # Try each role, starting with the one with the longest name.
+        # Otherwise we might see `@Fellow: Tuesday` and assume the mention was
+        # for the `@Fellow` role instead of the `@Fellow: Tuesday` role.
+        if message.include? "@#{role.name}"
+          # To mention a role in a Discord message, use the syntax:
+          # <@&ROLE_ID>.
```
Comment (sadleb) [inline app/services/schedule_discord_message.rb:55]: This comment is a bit confusing. If I understand correctly, to mention a role as the end user you just use the normal `@rolename` syntax, but internally we convert that to `<@&ROLE_ID>` for rendering purposes?

File: app/helpers/discord_schedule_helper.rb [hunk 5]
```diff
@@ -0,0 +1,18 @@
+module DiscordScheduleHelper
+  ROLE_MENTION_REGEX = /<@&(\d+)>/
```
Comment (sadleb) [inline app/helpers/discord_schedule_helper.rb:2]: It took me a minute to figure out what this was. Maybe add a comment here about how this is the format used in `convert_role_mentions`?

File: app/helpers/discord_schedule_helper.rb [hunk 6]
```diff
@@ -0,0 +1,18 @@
...
+  def render_discord_message(message)
+    # For now, this is just the inverse of
+    # ScheduleDiscordMessage.convert_role_mentions.
+    message.scan(ROLE_MENTION_REGEX).flatten.each do |discord_role_id|
+      # This is a role mention.
+      role = DiscordServerRole.find_by(discord_role_id: discord_role_id)
+      # Skip invalid role IDs, instead of crashing.
+      next unless role
+
+      message = message.sub("<@&#{discord_role_id}>", "@#{role.name}")
```
Comment (sadleb) [inline app/helpers/discord_schedule_helper.rb:13]: Is this when we're actually sending the message content to discord? Or is this just to show it in the platform UI? If it's us creating the message to send to discord, I don't get why we convert the original `@rolename` to `<@&ROLE_ID>`?

UPDATE: I googled and found out that `<@&ROLE_ID>` is actually the expected format for a role mention in a discord bot: https://stackoverflow.com/questions/52042695/discord-bot-role-mentioning. Maybe mention that somewhere and link to the docs?

File: spec/services/schedule_discord_message_spec.rb (hunk 1 above)
Comment (elenabaurkot) [reply spec/services/schedule_discord_message_spec.rb:8]: ah okay, so is '.run' the standard convention for when you describe actions in a service and '#run' is standard for controllers?

**Review by sadleb — COMMENTED** [review]
I eventually figured out what is going on here, so there's no need to respond to each comment but I'm leaving them there to highlight what was confusing for a first time reader. I think the key was just not knowing that `<@&ROLE_ID>` is the format expected by Discord when sending a message from a bot. I initially assumed you were just doing some validation / sanitization on the user input.

File: app/services/schedule_discord_message.rb (hunk 3 above)
Comment (rshipp) [reply app/services/schedule_discord_message.rb:49]: ah yeah that is better

File: spec/services/schedule_discord_message_spec.rb (hunk 1 above)
Comment (rshipp) [reply spec/services/schedule_discord_message_spec.rb:8]: if I'm describing a controller action, I use `#`. if I'm describing anything else, including a non-action method in a controller, I try to remember to use `.`.

Comment (sadleb) [reply spec/services/schedule_discord_message_spec.rb:8]: Actually, the convention is `#` for instance methods and `.` for class methods: https://www.betterspecs.org/

File: app/helpers/discord_schedule_helper.rb (hunk 2 above)
Comment (rshipp) [reply app/helpers/discord_schedule_helper.rb:11]: nah this isn't interesting, i just wanted to make sure entering weird stuff that accidentally matched the role-mention format wouldn't crash things.

File: spec/services/schedule_discord_message_spec.rb (hunk 1 above)
Comment (elenabaurkot) [reply spec/services/schedule_discord_message_spec.rb:8]: Good to know, thanks 👍

File: app/helpers/discord_schedule_helper.rb (hunk 6 above)
Comment (rshipp) [reply app/helpers/discord_schedule_helper.rb:15]: updated comments to hopefully clarify

**Review by sadleb — APPROVED** [review]

**Pull request merged**